File: java/src/test/java/io/github/nickliapin/tdc/generators/RegexSpaceTest.java
package io.github.nickliapin.tdc.generators;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * How many different strings a regex pattern makes, against the shared fixture.
 *
 * <p>{@code uniq="true"} over a {@code type="regex"} pattern refuses a count the pattern cannot
 * meet before it draws, and knows the pattern's size by counting its parse tree. That count
 * decides which configs are refused, so all five implementations are held to one set of numbers —
 * including the deliberate overcounts, which are the contract rather than a mistake.
 */
class RegexSpaceTest {

  private static JsonNode fixture() throws IOException {
    return new ObjectMapper()
        .readTree(Files.readString(Path.of("..", "fixtures", "cross-language", "regex-space.json")));
  }

  @Test
  @DisplayName("saturates at the same ceiling as the reference")
  void saturatesAtTheSameCeiling() throws IOException {
    assertEquals(fixture().get("cap").asLong(), RegexGen.SPACE_CAP);
  }

  @Test
  @DisplayName("counts every pattern's space the way the reference counts it")
  void countsTheSpaceTheReferenceCounts() throws IOException {
    for (JsonNode c : fixture().get("patterns")) {
      assertEquals(
          c.get("size").asLong(),
          RegexGen.spaceSize(c.get("pattern").asText(), RegexGen.DEFAULT_MAX_LENGTH),
          c.get("pattern").asText() + " — " + c.get("why").asText());
    }
  }

  @Test
  @DisplayName("counts every advanced pattern's space the way the reference counts it")
  void countsAnAdvancedPatternTheWayTheReferenceDoes() throws IOException {
    for (JsonNode c : fixture().get("advanced")) {
      assertEquals(
          c.get("size").asLong(),
          AdvancedRegexGen.spaceSize(c.get("pattern").asText(), RegexGen.DEFAULT_MAX_LENGTH),
          c.get("pattern").asText() + " — " + c.get("why").asText());
    }
  }
}

File: csharp/Tdcv2.Tests/RegexSpaceTest.cs
using System.IO;
using System.Text.Json;
using Tdcv2.Generators;
using Xunit;

namespace Tdcv2.Tests;

/// <summary>
/// How many different strings a regex pattern makes, against the shared fixture.
/// </summary>
/// <remarks>
/// <c>uniq="true"</c> over a <c>type="regex"</c> pattern refuses a count the pattern cannot meet
/// before it draws, and knows the pattern's size by counting its parse tree. That count decides
/// which configs are refused, so all five implementations are held to one set of numbers —
/// including the deliberate overcounts, which are the contract rather than a mistake.
/// </remarks>
public class RegexSpaceTest
{
    private static JsonDocument Fixture() =>
        JsonDocument.Parse(File.ReadAllText(
            Path.Combine(PrngVectorsTest.FixturesDir(), "regex-space.json")));

    [Fact]
    public void SaturatesAtTheSameCeiling()
    {
        using JsonDocument doc = Fixture();
        Assert.Equal(doc.RootElement.GetProperty("cap").GetInt64(), RegexGen.SpaceCap);
    }

    [Fact]
    public void CountsTheSpaceTheReferenceCounts()
    {
        using JsonDocument doc = Fixture();
        foreach (JsonElement c in doc.RootElement.GetProperty("patterns").EnumerateArray())
        {
            string pattern = c.GetProperty("pattern").GetString()!;
            long want = c.GetProperty("size").GetInt64();
            long got = RegexGen.SpaceSize(pattern, RegexGen.DefaultMaxLength);
            Assert.True(
                want == got,
                $"{pattern}: expected {want}, counted {got} — {c.GetProperty("why").GetString()}");
        }
    }
}
